### ISSUE TITLE
unknown error for NL 6.1, not a problem for previous versions

### ISSUE THREAD

**Issue by cmbarton** [NetLogo/NetLogo#1763]
This happens in 6.1 but did not in prior versions (<= 6.0.4). A call to "ask" a turtle-set is highlighted in blue but not a yellow error highlight. Running on Mac OS X 10.14.5

ERROR: 

NetLogo is unable to supply you with more details about this error.  Please report the problem 
 at https://github.com/NetLogo/NetLogo/issues, or to bugs@ccl.northwestern.edu, and paste the 
contents of this window into your report. 

java.lang.ArrayIndexOutOfBoundsException
 at org.nlogo.agent.InRadiusOrCone.setPatches(InRadiusOrCone.scala:397)
 at org.nlogo.agent.InRadiusOrCone.inRadius(InRadiusOrCone.scala:43)
 at org.nlogo.prim._asm_proceduremove_ifelse_13.perform(:3)
 at org.nlogo.nvm.Context.runExclusive(Context.java:133)
 at org.nlogo.nvm.ExclusiveJob.run(ExclusiveJob.scala:39)
 at org.nlogo.nvm.Context.runExclusiveJob(Context.java:177)
 at org.nlogo.prim._hatch.perform(_hatch.java:60)
 at org.nlogo.nvm.Context.runExclusive(Context.java:133)
 at org.nlogo.nvm.ExclusiveJob.run(ExclusiveJob.scala:39)
 at org.nlogo.nvm.Context.runExclusiveJob(Context.java:177)
 at org.nlogo.prim._asm_procedurevillagefission_ask_0.perform(:1)
 at org.nlogo.nvm.Context.stepConcurrent(Context.java:107)
 at org.nlogo.nvm.ConcurrentJob.step(ConcurrentJob.scala:65)
 at org.nlogo.job.JobThread.runPrimaryJobs(JobThread.scala:133)
 at org.nlogo.job.JobThread.$anonfun$run$1(JobThread.scala:68)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
 at scala.util.control.Exception$Catch.apply(Exception.scala:228)
 at org.nlogo.api.Exceptions$.handling(Exceptions.scala:41)
 at org.nlogo.job.JobThread.run(JobThread.scala:66)

NetLogo 6.1.0
main: org.nlogo.app.AppFrame
thread: JobThread
Java HotSpot(TM) 64-Bit Server VM 1.8.0_202 (Oracle Corporation; 1.8.0_202-b08)
operating system: Mac OS X 10.14.5 (x86_64 processor)
Scala version 2.12.8
JOGL: (3D View not initialized)
OpenGL Graphics: (3D View not initialized)
model: village_farming3

02:52:52.167 JobRemovedEvent (org.nlogo.app.App$$anon$4 (org.nlogo.window.GUIWorkspace)) JobThread
02:52:52.166 SwitchedTabsEvent (org.nlogo.app.Tabs) AWT-EventQueue-0
02:52:52.122 RuntimeErrorEvent (org.nlogo.app.App$$anon$4 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
02:52:51.944 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.944 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.943 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.943 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.943 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.943 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
02:52:51.943 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0

**Comment by LaCuneta** [comment]
Michael,

Thank you for reporting this issue, we appreciate it, and sorry you ran into it!

I confirmed this is a regression in NetLogo 6.1.0 from 6.0.4.  It occurs when an `in-radius` command is run, then the world size changes somehow, then another `in-radius` command is executed.  The world size change could be a change to a model's view, or possibly loading a different model with a differently sized world.  A minimum set of code to reproduce is then:

```
crt 2 [ fd 100 ]
ask turtle 0 [ ask turtles in-radius 100 [ set color blue ] ]
resize-world -50 50 -50 50
__change-topology false false
crt 2 [ fd 100 ]
ask turtle 0 [ ask turtles in-radius 100 [ set color blue ] ]
```

There is fortunately a workaround.  If you run into the error, then save your model (if you have any changes), close NetLogo entirely (not just your model), and re-open NetLogo and your model and you should not see the exception when you run.  Closing and re-opening gets rid of the "bad state" that causes the error.

I'll work on a fix for this, but let me know if you have any questions on it, or if the scenario I outlined doesn't sound similar to how you encountered the error.

**Comment by UriCCL** [comment]
Can we touch base today?

-- Uri

**Comment by cmbarton** [comment]
I'm in Europe. So there is some time delay in our communication. It does seem to be bombing at an in-radius call. But I do not change the world in any way, anywhere in the model. Also the work around does not work. I can quit NetLogo and try again. But it always bombs. Interestingly, it will run >100 iterations before finally bombing. So this seems to be variant behavior for this bug with no work around.

**Comment by cmbarton** [comment]
Also, I'm happy to test a bug fix but don't have experience compiling Java.

**Comment by LaCuneta** [comment]
@cmbarton Thanks for the extra details.  

Do you have the exact steps you're following to cause the error?  I'm fairly certain it's the same issue I found since the locations in code are identical in the exception message, but it would be good to be certain.  It could happen even when your model just has a different world size than the default, since the blank model usually loads when you open NetLogo.

If it's happening only in your village_farming3 model, would you mind posting the code for that model here, or if you don't want it in public sending it in to bugs@ccl.northwestern.edu?

**Comment by cmbarton** [comment]
Hi Jeremy,

I'm sending my model code to the bugs email. I'm doing nothing to cause the error. Just running the model (getting ready for a presentation) and eventually it bombs. It doesn't bomb the first time the move procedure is called. Seems to be kind of random as to when it happens. There are no calls in the model to resize the world. Maybe you can figure it out.

**Comment by TheBizzle** [comment]
@cmbarton, I understand your last message to mean that you already sent your model to bugs@ccl.northwestern.edu (probably yesterday), but we didn't receive it.  Please ensure that you sent it to the right address and try again.

**Comment by cmbarton** [comment]
Jeremy,

I just sent the file again to bugs@ccl.northwestern.edu. This time I zipped the file first, because some university spam filters stop mail with *.nlogo attachments.

Michael

**Comment by TheBizzle** [comment]
Great.  We received the file this time, and I can confirm that I'm running into the error that you described.  We'll investigate more and get back to you here.

**Comment by cmbarton** [comment]
Thanks for the update 

Michael Barton

...Sent from my iPad

**Comment by TheBizzle** [comment]
I've minimized the test case and confirmed that it throws an error even with the latest version of the `hexy` branch.

Steps to reproduce:

  * Open a new model
  * Ensure that the world wraps in *both* directions and the world origin is "Center"
  * Set `max-pxcor` to `60`
  * Set `max-pycor` to `50`
  * Set the model's code to the following and then run `setup`:

```nlogo
to setup
  clear-all
  crt 1 [ setxy -35 -45 ]
  crt 1 [ setxy -23  16 ]
  ask turtle 1 [ let boomer patches in-radius 55 ]
end
```

Back to you, @LaCuneta!

**Comment by LaCuneta** [comment]
@cmbarton Thanks again for submitting the model it's what we needed to find the root cause.  I've got a fix running through tests now that I've confirmed resolves the issue with your model.

In the meantime, the best workaround is going to be using the `temp-in-radius` custom procedure I suggested in [this issue](https://github.com/NetLogo/NetLogo/issues/1756).  It's not too hard to implement, but it could slow down your model a bit.  The next best option is to revert to 6.0.4 if you're not using any of the 6.1.0 features, until we get a 6.1.1 fix release out. 

Let me know if you have any questions.
